shark_turbine/kernel/wave/expansion.py

```diff
@@ -129,15 +147,42 @@ def set_node_index(
                 and isinstance(custom, MMA)
             )
 
+            vector_check = (
+                isinstance(constraint, HardwareConstraint)
+                and constraint.vector_shapes is not None
+                and hasattr(custom, "elements_per_thread")
+            )
+
             constraint_check = (
                 not isinstance(constraint, HardwareConstraint) and dim == constraint.dim
             )
 
-            if (not mma_check) and (not constraint_check):
+            if (not (mma_check or vector_check)) and (not constraint_check):
                 continue
 
             if isinstance(constraint, HardwareConstraint):
-                index_seq = constraint.apply(mma_index[dim])
+
+                # The semantics of elements_per_thread are that it represents the number of
+                # elements that are loaded contiguously from memory.
+                elements_per_thread = getattr(custom, "elements_per_thread", None)
+                constraint_index, elements_per_thread = (
+                    (
+                        workgroup_constraints[dim].workgroup_dim,
+                        (
+                            0
+                            if not is_contiguous_dim(
+                                dim,
```

Comment:
Do we expect the indexExpr to be `0, 0, 1` after this or should we just raise error?

Reply:
So this is why I added the comment about the semantics of elements_per_thread. What it specifies is the number of elements that can be read contiguously by a single thread. But when you specify that on a read, the read could have a type of [M, N]. So what this function does is specify the access pattern of N to be consistent with elements_per_thread and sets the access_pattern of M to be 0 elements. So we should not raise an error in this case, it is the expected access pattern along the M dimension.

Reply:
Understood, thanks for the explanation! :)

May or may not mean anything but shouldn't the generated IndexSequence here be 0,1,1? since there is 1 element there?

Reply:
So that's an interesting question. When you are loading a 1d vector from dimension N of a [M, N] tensor then are you loading [0, 4] elements or [1, 4] elements ? Basically we can represent a 1d vector as vector<4xf32> (which corresponds to 0 elements along the M dimension) or as vector<1x4xf32> (which corresponds to 1 element along the M dimension). They both represent the same quantity, so makes sense to go with the most concise representation which corresponds to M = 0 (vector<4xf32>).

Reply:
Ah OK, that makes sense, thanks! :)